### PR TITLE
Use median time for DPoS consensus

### PR DESCRIPTION
**Pull request by RainFallsSilent**

### CODE REVIEW

File: dpos/manager/dposmanager.go [hunk 1]
```diff
@@ -340,14 +345,89 @@ func (d *DPOSManager) recoverAbnormalState(firstRecover bool) bool {
...
+
+func medianOf(nums []int64) int64 {
+	l := len(nums)
+
+	if l == 0 {
+		return 0
+	}
+
+	if l%2 == 0 {
+		return nums[l/2] + nums[l/2-1]/2.0
```
Comment (alexpan0610) [inline dpos/manager/dposmanager.go]: (nums[l/2] + nums[l/2-1])/2.0

File: dpos/manager/dposmanager.go [hunk 2]
```diff
@@ -340,14 +351,89 @@ func (d *DPOSManager) recoverAbnormalState(firstRecover bool) bool {
...
+			status = v
+		}
+		startTimes = append(startTimes, v.ViewStartTime.UnixNano())
+	}
+	sort.Slice(startTimes, func(i, j int) bool {
+		return startTimes[i] < startTimes[j]
+	})
+	medianTime := medianOf(startTimes)
+	sec := medianTime / 1e9
+	nsec := medianTime - sec
```
Comment (alexpan0610) [inline dpos/manager/dposmanager.go]: Wrong math! should be nsec := medianTime - sec*1e9.

File: dpos/manager/dposmanager.go [hunk 3]
```diff
@@ -340,14 +351,89 @@ func (d *DPOSManager) recoverAbnormalState(firstRecover bool) bool {
...
+	status.ViewStartTime = time.Unix(sec, nsec)
```
Comment (alexpan0610) [inline dpos/manager/dposmanager.go]: Use status.ViewStartTime = dtime.Int64ToTime(medianTime) instead.

File: dpos/manager/dposmanager.go (hunk 1 above)
Comment (RainFallsSilent) [reply dpos/manager/dposmanager.go]: fixed

File: dpos/manager/dposmanager.go (hunk 2 above)
Comment (RainFallsSilent) [reply dpos/manager/dposmanager.go]: fixed.....

File: cmd/script/api/dposmanager.go [hunk 4]
```diff
@@ -3,6 +3,7 @@ package api
 import (
 	"encoding/json"
 	"fmt"
+	"github.com/elastos/Elastos.ELA/dpos/dtime"
```
Comment (alexpan0610) [inline cmd/script/api/dposmanager.go]: Format

Comment (RainFallsSilent) [reply cmd/script/api/dposmanager.go]: fixed

**Pull request merged**